README.md
# PLUG SDK

PLUG SDK는 모바일 게임 사용자가 게임 도중 게임을 벗어나지 않고 게임의 커뮤니티와 소통할 수 있도록 제공되는 인앱 커뮤니티 라이브러리입니다. 간단한 조작으로 프로젝트에 PLUG SDK를 추가할 수 있습니다.

이미 많은 게임에 적용돼 검증된 PLUG SDK를 사용하면 손쉽게 커뮤니티 기능을 게임 안에 넣을 수 있습니다.
 
![PLUG SDK](http://static.naver.net/m/cafe/glink/promotion/cafe_sdk_open/img_intro1_20151111.png)

## How To Get Started 

### 제휴 신청

PLUG SDK를 앱에 적용하려면 PLUG SDK 제휴를 신청해야 합니다. 제휴 양식을 작성해 PLUG 제휴 담당자에게 메일로 제휴를 신청합니다.

- [제휴 양식 다운로드](https://github.com/naver/cafe-sdk-android/raw/master/guide/alliance/%EB%84%A4%EC%9D%B4%EB%B2%84%EC%B9%B4%ED%8E%98SDK_%EC%A0%9C%ED%9C%B4%EC%96%91%EC%8B%9D_%EA%B2%8C%EC%9E%84%EC%82%AC%EB%AA%85_%EA%B2%8C%EC%9E%84%EB%AA%85_ver.2.0.0.xlsx)
- PLUG 제휴 담당자 메일 주소: <a href="mailto:dl_naverplugpartner@navercorp.com">dl_naverplugpartner@navercorp.com</a>

제휴 신청이 완료되면 PC의 카페 관리 페이지에서 네이버 PLUG SDK 콘텐츠를 관리할 수 있습니다.

### 적용 가이드

- [PLUG 적용 가이드](https://www.gitbook.com/book/plug/plug-sdk-ios/details)
- [최신 버전 및 샘플 프로젝트 다운로드](https://github.com/naver/cafe-sdk-ios/archive/master.zip)

## Communication 

개발과 관련된 문의 사항은 PLUG 공식 카페를 사용합니다.

- [PLUG 공식 카페](http://cafe.naver.com/navercafesdk)
- [PLUG 공식 카페의 개발 문의 게시판](http://cafe.naver.com/ArticleList.nhn?search.clubid=28285034&search.menuid=13&search.boardtype=L)

## Usages

### 1. 요구 사항

- 애플리케이션 사용 환경: iOS 9.0 이상(64비트 아키텍처 지원, bitcode 지원)
- 동영상 녹화: iOS 9.0 이상, A7 이상
- 실시간 스트리밍: iOS 10.0 이상, A7 이상
- IDE: XCode 9.0 이상

### 2. 프로젝트 설정

PLUG SDK를 사용하려면 XCode에서 다음과 같이 개발 환경을 설정합니다.

1. **Build Settings**의 **Other Linker Flags**에 **-ObjC** 옵션을 추가합니다.  
2. 다운로드한 [샘플 프로젝트](https://github.com/naver/cafe-sdk-ios/archive/master.zip)에 포함된 다음 라이브러리를 프로젝트에 추가합니다.    
    - SDWebImage  
    - FLAnimatedImage  
    - AFNetworking  
    - 네이버 아이디로 로그인
    - OpenSSL
3. **Build Phases**의 **Link Binary With Libraries**에 다음 라이브러리를 추가합니다. 
    - AVKit.framework
    - AVFoundation.framework
    - MobileCoreServices.framework
    - SystemConfiguration.framework
    - MediaPlayer.framework
    - AVFoundation.framework
    - CoreMedia.framework
    - AssetsLibrary.framework
    - Security.framework
    - ImageIO.framework
    - QuartzCore.framework
    - WebKit.framework
    - ReplayKit.framework
4. **Info.plist**의 사용자 권한 획득 옵션을 추가합니다.
    - NSCameraUsageDescription
    - NSPhotoLibraryUsageDescription
    
### 3. 초기화

PLUG SDK가 지원하는 언어에 따라 다음과 같이 PLUG SDK를 초기화합니다.

#### 3.1 PLUG SDK 초기화(한국어만 지원 - 네이버 카페)

한국어만 지원하는 PLUG SDK를 초기화할 때는 다음과 같은 초기화 정보를 설정합니다.

- [네이버 아이디로 로그인 개발자 센터](https://developers.naver.com/apps/#/register?api=nvlogin)에서 애플리케이션을 등록하고 받은 클라이언트 아이디(ClientId)와 클라이언트 시크릿(ClientSecret)
- [네이버 카페](http://section.cafe.naver.com/)에서 카페를 개설하고 받은 카페 ID
- [네이버 아이디로 로그인 개발자 센터](https://developers.naver.com/apps/#/register?api=nvlogin)에 애플리케이션을 등록할 때 입력한 [URL Scheme](https://developers.naver.com/apps/#/myapps)
- PLUG가 올라갈 부모 뷰 컨트롤러
- 채널 코드(한국어 채널만 사용할 때는 `KOREAN`으로 설정)

```objective-c
    [[NCSDKManager getSharedInstance] setNaverLoginClientId:@"197CymaStozo7X5r2qR5"
                                     naverLoginClientSecret:@"evCgKH1kJL"
                                                     cafeId:28290504];
                                                     
    [[NCSDKLoginManager getSharedInstance] setNaverLoginURLScheme:@"gLinkSample"];
    
    [[NCSDKManager getSharedInstance] setParentViewController:self];
    
    [[NCSDKManager getSharedInstance] setChannelCode:KOREAN];
```

#### 3.2 PLUG SDK 초기화(외국어만 지원)

외국어만 지원하는 PLUG SDK를 초기화할 때는 다음과 같은 초기화 정보를 설정합니다.

- consumerKey: 커뮤니티를 개설하고 받은 컨슈머 키(Consumer Key)
- consumerSecretKey: 커뮤니티를 개설하고 받은 컨슈머 시크릿(Consumer Secret Key)
- cummunityNo: 커뮤니티를 개설하고 받은 커뮤니티 No(cummunityNo)
- loungeNo: 라운지를 개설하고 받은 라운지 No(cummunityNo)
- PLUG가 올라갈 부모 뷰 컨트롤러

```objective-c
    [[NCSDKManager getSharedInstance] setGlobalConsumerKey:@"PLUGTESTKEY"
                                      globalConsumerSecret:@"PLUGTESTSECRET" 
                                      globalCommunityNo:1
                                      globalLoungeNo:58];
                                              
    [[NCSDKManager getSharedInstance] setParentViewController:self];
```

> **커뮤니티 ID와 컨슈머 키 확인**  
> 커뮤니티를 개설하고 받은 커뮤니티 ID와 컨슈머 키를 확인하려면 [커뮤니티 관리 페이지](http://g.cafe.naver.com/plugsample/manage/consumer)에 매니저 권한으로 접근하십시오.  
> 커뮤니티 관리 페이지에서 **Settings > Communities > In-App Community Settings** 메뉴를 선택하면 커뮤니티 ID와 컨슈머 키를 확인할 수 있습니다.   
> ![In-App Community Settings](https://plug.gitbooks.io/plug-sdk-android/content/assets/wiki-plug-setting.png)

### 4. PLUG SDK 시작

초기화가 완료되면 PLUG SDK를 실행합니다.

#### 4.1 presentMainViewController: 메서드

presentMainViewController: 메서드는 PLUG SDK를 시작하는 메서드입니다. 다음은 presentMainViewController: 메서드를 구현한 예입니다.

```objective-c
[[NCSDKManager getSharedInstance] presentMainViewController];

```

> **NCSDKManager 클래스**  
> NCSDKManager 클래스는 PLUG SDK를 제어하는 클래스입니다. NCSDKManager 클래스로 제어할 수 있는 기능에 관한 자세한 내용은 PLUG SDK 상세 가이드의 "[주요 기능 설명 상세](https://plug.gitbooks.io/plug-sdk-ios/content/ko/index.html)"를 참고하십시오.

#### 4.2 가로 모드, 세로 모드

PLUG SDK는 가로 모드와 세로 모드를 모두 지원합니다. 다음은 PLUG SDK를 가로 모드로 실행하는 예입니다.

```objective-c
[[NCSDKManager getSharedInstance] setOrientationIsLandscape:YES];

```

## Files and Folders

|폴더 및 파일 이름	|설명|
|---|---|
|guide/	|PLUG SDK 적용 가이드가 있는 폴더
|lib/	|iOS용 PLUG SDK 라이브러리 파일이 있는 폴더
|sample/external-lib	|외부 라이브러리 파일이 있는 폴더|
|sample/navercafesdk-sample-ios	|Xcode용 샘플 프로젝트가 있는 폴더|
|LICENSE 	|라이선스 파일|
|README.md	|리드미 파일|

## Dependent Libraries

- PLUG SDK를 사용하려면 다음 라이브러리를 프로젝트에 추가해 함께 빌드해야 합니다. 라이브러리는 샘플 프로젝트에 포함돼 있습니다.[네이버 아이디 로그인](https://nid.naver.com/devcenter/docs.nhn?menu=IOS)(4.0.7 버전)
- [AFNetworking](https://github.com/AFNetworking/AFNetworking)(3.1.0 버전 )
  - AFNetworking 2.x 버전과 3.x 버전 모두 사용할 수 있습니다. PLUG SDK 테스트는 AFNetworking 3.1.0 버전으로 진행했습니다.

- [SDWebImage](https://github.com/rs/SDWebImage)(3.8.2 버전)
- [FLAnimatedImage-](https://github.com/Flipboard/FLAnimatedImage)(1.0.12 버전)
- ApiGateway-1.5.0
- KaleidoCore.framework
  
## License 

PLUG SDK의 저작권은 네이버(주)에 있습니다.

```
NAVER Cafe SDK for iOS

Copyright 2015 NAVER Corp.
All rights reserved.

Unauthorized use, modification and redistribution of this software are strongly prohibited.
```

PLUG SDK 3.x 버전은 더 이상 지원되지 않습니다.



CHANGELOG.md
v.4.4.7
-------------
### fixed
1. iOS SDK 처음 실행 중, 게시글 이동 기능이 동작하지 않는 이슈 수정.
2. iOS 외부 라이브라리 SDWebImage 4 버전 사용 시, 빌드 오류 수정.
3. iOS 샘플 프로젝트 NaverLogin 모듈 4.1.1 버전 업데이트.

v.4.4.6
-------------
### fixed
1. iOS SDK 처음 실행 시, 게시글 이동 기능이 동작하지 않는 이슈 수정.
2. iOS ShowWidgetWhenUnloadSDK를 NO 설정 시, 최소화 버튼 숨김이 되지 않는 이슈 수정.
3. iOS Xcode 10 버전에서 빌드 시 오류 발생 이슈 수정.

v.4.4.5
-------------
### fixed
1. Rebuild with Xcode 11.2.1
Storyboards containing a UITextView will cause the app to crash on operating system versions earlier than iOS 13.2 and tvOS 13.2 if compiled with Xcode 11.2. (56808566)
https://forums.developer.apple.com/thread/125287
https://developer.apple.com/documentation/xcode_release_notes/xcode_11_2_1_release_notes

v.4.4.4 (Deprecated)
-------------
### fixed
1. iOS UIWebView -> WKWebView 교체

v.4.4.3
-------------
### fixed
1. iOS Moot COPPA 대응

v.4.4.2
-------------
### fixed
1. iOS 13 대응

v.4.4.1
-------------
### fixed
1. iOS AFNetworking Framework 의존성 제거.
2. iOS ApiGateway Framework 의존성 제거.
3. iOS setChannelCode 설정 오류 수정.
4. iOS App Bundle Preferred Localization 기준으로 채널 설정되게 수정.

v.4.3.1
-------------
### fixed
1. iOS OpenSSL Framework 의존성 제거.

v.4.3.0
-------------
### fixed
1. 간헐적으로 발생하는 Major 버그 수정

v.4.2.2
-------------
### fixed
1. Unity 백그라운드 화면 터치 이슈 수정

v.4.2.1
-------------
### updated
1. PLUG 시작시 특정 메뉴를 선택할 수 있도록 하는 기능 추가

### fixed
1. iPad UI 버그 수정( iOS )
2. 게시글 바로가기 Interface 버그 수정
4. 기타 알려진 버그 수정


v.4.2.0
-------------
### updated
1.  PLUG 최적화
2.  Live 제거


v.4.1.1
-------------
### fixed
1. 위젯터치 오류 수정 ( iOS )

v.4.1.0
-------------
### fixed
1. PLUG 배너에서 "외부링크" 항목 추가
2. 기타 알려진 버그 수정

v.4.0.3
-------------
### updated
1. Fixed iPhone 5 Crash

v.4.0.2
-------------
### updated
1. iOS에서 Global 만 설정하는 경우 크래시 발생 수정.

v.4.0.0
-------------
### updated
1. Moot 연동

v.3.3.1
-------------
### updated
1. showWidgetWhenUnloadSDK 를 사용할 경우 최소화버튼이 보이지 않도록 수정


v.3.3.0
-------------
### updated
1. 채널변경 on/off 인터페이스 추가
2. KaleidoCore.framework 를 static library  로 변경

### fixed
1. 알려진 버그 수정


v.3.2.6
-------------
### fixed
1. 앱 Window 사이즈가 Screen 사이즈랑 맞지 않을 화면 깨지는 문제 수정.
2. 세로모드 분할형 배너 동영상에서 화면 깨지는 문제 수정.

v.3.2.4
-------------
### fixed
1. KaleidoCore.framework 1.4.11 업데이트
2. 라이브중 상대방 위젯에서 live 표시가 잘못 노출되는 이슈 수정

v.3.2.3
-------------
### fixed
1. 플러그 실행 중 게임으로 이벤트 넘어가는 현상 수정
2 . 배너 홈에서 메인 배너 영역이 정사각형으로 노출되지 않는 것 수정.

v.3.2.2
-------------
### fixed
1. 한국어 이외의 언어에서 전체 게시글, 오늘 방문자, 외 X명 번역어 잘못 나오는 이슈 수정

v.3.2.1
-------------
### fixed
1. 특정 단말에서 위젯 버튼 클릭시 크래시 이슈 수정 (아이폰 5c)

v.3.2.0
-------------
### updated
1. 라이브탭 추가
2. 위젯 노출 버튼 변경
3. 라이브 VOD 정렬 기능 추가
4. 좋아요 로직 수정
5. 아이폰 X Safe Area UX 대응

v.3.1.2
-------------
### fixed
1. 누락된 번역 문구 추가

v.3.1.1
-------------
### fixed
1. 라이브/VOD 시청 콜백 오류 수정
2. 네이버 프로필 API 오류 수정 (NCNaverLoginManager.getNaverIdProfile)

v.3.1.0
-------------
### updated
1. 라이브 기능 추가 (NaverCafeSDKLive.framework)
2. 디자인 변경
3. 인기글 적용
4. 글로벌 커뮤니티 지역별 관리 기능 추가 (http://cafe.naver.com/navercafesdk/1550)
5. 필독공지 노출 개수 증가 (3개)
6. @3x 이미지 적용
7. 위젯 위치 수정 (iPhone X Notch 회피)

v.2.5.5
-------------
### updated
1. XCode 9 & iPhone X 대응

v.2.5.2
-------------
### updated
1. iOS 11 대응

v.2.5.1
-------------
### updated
1. 동영상 녹화 기능 제거 버전 (ReplayKit.framework 참조 제거)

v.2.5.0
-------------
### updated
1. 본분/댓글 번역기능 추가
2. 지원 언어 추가 (브라질, 포르투갈, 이탈리아, 멕시코, 터키, 베트남)
3. 녹화 인터페이스 제공
4. 네이버 로그인 라이브러리 사용시 최초 세팅한 URLScheme, delegate 갱신하도록 변경

v.2.4.5
-------------
### updated
1. 중국어 번체/간체 오류 수정

v.2.4.3
-------------
### updated
1. 러이아어 추가
2. 동영상 업로드 시 크기에 따른 압축 품질 변경 로직 추가

v.2.4.0
-------------
### updated
1. Pinned 공지 기능 추가
2. 댓글 목록에 답글 3개 미리 보기 추가
3. WKWebView 적용 (iOS)
4. GLReachability 클래스명 변경 (iOS)
5. SDK 홈 배너 클릭시 발생하는 이벤트를 delegate로도 처리되도록 추가 (ncSDKAppSchemeBanner:)
* 기존에는 AppDelegate openURL을 호출

### fixed
1. 알려진 이슈 수정

v.2.3.1
-------------
### fixed
1. 위젯 사라지는 현상 수정
2. 댓글 델리게이트 호출 안되는 현상 수정

v.2.3.0
-------------
### updated
1. 동영상/이미지 모아보기 기능 추가
2. 접기/닫기 기능 개선
3. 위젯 기능 개편
4. 비공개 카페 테스트 가능하도록 변경
5. 스페인어/프랑스어/독일어/인도네시아어 추가
6. 네이버 로그인 관련 API 추가
    1) NCNaverLoginManager.h
7. API 변경
    1) presentArticlePostViewController 관련 파람 변경
    2) setCountry -> setChannelCode
8. API 추가
    2) setWidgetStartPosition
9. AFNetworking-3.1.0 적용

### fixed
1. 알려진 이슈 수정

v.2.2.0
-------------
####updated####
1. PLUG에 테마 color 적용 가능
2. PLUG 접기 기능 추가
3. 프로필 사진 업데이트 기능 추가
4. default channel이 device 지정 언어에 따라가도록 변경

v.2.1.0
-------------
####updated
1. 세로모드 지원


v.2.0.1
-------------
####fixed####
1. 글로벌 세팅 후 한국으로 로드 했을때 채널변경 버튼 나오지 않는 이슈 수정


v.2.0.0
-------------
####updated
1. 영어, 중국어, 일본어, 태국어 채널 지원.

####fixed####
1. iOS 10 관련 이슈 수정.

####issue####
1. 세로모드 2.1.0에서 지원 예정.
2. 언어 설정 중국어인 경우 위젯 동영상 녹화 기능 안되는 이슈.

v.1.7.0
-------------
####updated
1. 위젯을 통하여 동영상 녹확 기능을 제공합니다.
2. 콜백 API 개선 (투표 완료 콜백을 제공합니다)


####fixed####
1. CFBundleSupportedPlatforms 이슈 수정


v.1.6.0
-------------
####updated
1. 카페 위젯을 지원하여 게임내 어디서든 카페를 이용 할 수 있습니다.


####fixed
1. 게시글 간 이동이 자연스럽게 보이도록 수정


v.1.5.0
-------------
####updated
1. 타인의 프로필을 열람할 수 있습니다.

####fixed
1. 알려진 버그 수정


v.1.4.0
-------------
####updated
1. 프로필 화면에서 작성한 글, 댓글단 글, 좋아요 한 글을 볼 수 있습니다.
2. 카페 SDK 로고를 터치하면 카페 SDK 버전을 확인 할 수 있습니다
3. 글쓰기, 댓글쓰기, 카페가입 리스너를 등록하여 이벤트를 받아 사용할 수 있습니다.

####fixed
1. 알려진 버그 수정


v.1.3.0
-------------
####updated
1. 이미지를 클릭하여 크게 보기를 지원합니다.
2. 특정 게시판으로 바로 이동 할 수 있는 기능을 지원합니다.
3. 투명도 조절바를 on/off 할 수가 있습니다.

####fixed
1. 알려진 버그 수정


v.1.2.0
-------------
####updated
1. 투명도 조절 지원
2. 최신글 목록 보기 지원
3. 게시글 바로가기 기능 지원


####fixed
1. 알려진 버그 수정


v.1.1.0
-------------
###updated
1. 세로형 app 지원. (세로형 UI가 새로 적용되었습니다)
2. 전체 게시판 검색 지원.
3. 카페SDK 시작 및 종료 리스너 추가
    (setOnSdkStartedListener() 와 setOnSdkStoppedListener() 를 통하여 제공)


v.1.0.0
-------------
 NaverCafeSDK 오픈


sample/navercafesdk-sample-ios/navercafesdk-sample/AppDelegate.m
//
//  AppDelegate.m
//  navercafesdk-sample
//
//  Created by Gyeonghwan on 2015. 10. 26..
//  Copyright (c) 2015년 navercorp. All rights reserved.
//

#import "AppDelegate.h"
#import <NaverCafeSDK/NCSDKLoginManager.h>
@interface AppDelegate ()

@end

@implementation AppDelegate


- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    // Override point for customization after application launch.
    return YES;
}

- (void)applicationWillResignActive:(UIApplication *)application {
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application {
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application {
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application {
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application {
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url sourceApplication:(NSString *)sourceApplication annotation:(id)annotation {
    [[NCSDKLoginManager getSharedInstance] finishNaverLoginWithURL:url];
    return YES;
}
- (UIInterfaceOrientationMask)application:(UIApplication *)application supportedInterfaceOrientationsForWindow:(UIWindow *)window {
    return UIInterfaceOrientationMaskAll;
}
@end


sample/navercafesdk-sample-ios/navercafesdk-sampleTests/navercafesdk_sampleTests.m
//
//  navercafesdk_sampleTests.m
//  navercafesdk-sampleTests
//
//  Created by Gyeonghwan on 2015. 10. 26..
//  Copyright (c) 2015년 navercorp. All rights reserved.
//

#import <UIKit/UIKit.h>
#import <XCTest/XCTest.h>

@interface navercafesdk_sampleTests : XCTestCase

@end

@implementation navercafesdk_sampleTests

- (void)setUp {
    [super setUp];
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

- (void)testExample {
    // This is an example of a functional test case.
    XCTAssert(YES, @"Pass");
}

- (void)testPerformanceExample {
    // This is an example of a performance test case.
    [self measureBlock:^{
        // Put the code you want to measure the time of here.
    }];
}

@end


sample/external-lib/NaverLogin-thirdPartyModule/NLoginThirdPartyOAuth20InAppBrowserViewController.h
//
//  NLoginThirdPartyOAuth20InAppBrowserViewController.h
//  NaverOAuthSample
//
//  Created by TY Kim on 2014. 8. 20..
//  Copyright 2014 Naver Corp. All rights reserved.
//

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>
#import "NaverThirdPartyLoginConnection.h"

@interface NLoginThirdPartyOAuth20InAppBrowserViewController : UIViewController <UIWebViewDelegate> {
    NaverThirdPartyLoginConnection *_thirdPartyLoginConn;
    
    NSString *_linkUrl;
    NSURLRequest *_urlRequest;
    
    UIView *_mainView;
    
    UIView *_bannerView;
    UIButton *_bannerCloseButton;
    UIWebView *_webView;
	BOOL _isLeaving;
    BOOL _isCloseBannerView;
    
    UIView *_bottomBar;
    UIView *_diviView;
    UIButton *_closeBt;
    
    int _topMargin;

    
    UIInterfaceOrientation parentOrientation;
}

@property (nonatomic, assign) UIInterfaceOrientation parentOrientation;

- (id) initWithRequest:(NSURLRequest *) urlRequest;
- (void)setOAuthWebViewDelegate:(id)aDelegate;
- (NSString *)parameterValueWithUrl:(NSString *)url paramName:(NSString *)name;

@end


sample/navercafesdk-sample-ios/navercafesdk-sample/ViewController.m
//
//  ViewController.m
//  navercafesdk-sample
//
//  Created by Gyeonghwan on 2015. 10. 26..
//  Copyright (c) 2015년 navercorp. All rights reserved.
//

#import "ViewController.h"
#import <NaverCafeSDK/NCSDKManager.h>
@interface ViewController () <NCSDKManagerDelegate, UIAlertViewDelegate>

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];

    // Do any additional setup after loading the view, typically from a nib.
    
    //use Naver Cafe (korea)
    [[NCSDKManager getSharedInstance] setNaverLoginClientId:@"197CymaStozo7X5r2qR5"
                                     naverLoginClientSecret:@"evCgKH1kJL"
                                                     cafeId:28290504];
    //use Moot (global)
    [[NCSDKManager getSharedInstance] setGlobalConsumerKey:@"PLUGTESTKEY"
                                      globalConsumerSecret:@"PLUGTESTSECRET"
                                         globalCommunityNo:1
                                            globalLoungeNo:58];

    [[NCSDKManager getSharedInstance] setParentViewController:self];
    [[NCSDKManager getSharedInstance] setNcSDKDelegate:self];
    [[NCSDKManager getSharedInstance] setUseWidgetScreenShot:YES];
    [[NCSDKManager getSharedInstance] setUseWidgetVideoRecord:YES];
    
    UIButton *button1 = [[UIButton alloc] initWithFrame:CGRectMake(10, 80, 40, 40)];
    [button1 setImage:[UIImage imageNamed:@"icon1.png"] forState:UIControlStateNormal];
    [button1 addTarget:self action:@selector(touchButton1) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:button1];
    
    UIButton *button2 = [[UIButton alloc] initWithFrame:CGRectMake(10, CGRectGetMaxY(button1.frame), 40, 40)];
    [button2 setImage:[UIImage imageNamed:@"icon2.png"] forState:UIControlStateNormal];
    [button2 addTarget:self action:@selector(touchButton2) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:button2];

    UIButton *button3 = [[UIButton alloc] initWithFrame:CGRectMake(10, CGRectGetMaxY(button2.frame), 40, 40)];
    [button3 setImage:[UIImage imageNamed:@"icon3.png"] forState:UIControlStateNormal];
    [button3 addTarget:self action:@selector(touchButton3) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:button3];
}
- (void)setOrientationMode {
    [[NCSDKManager getSharedInstance] setOrientationIsLandscape:UIInterfaceOrientationIsLandscape([UIApplication sharedApplication].statusBarOrientation)];
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)touchButton1 {
    [self setOrientationMode];
    [[NCSDKManager getSharedInstance] presentMainViewController];
}
- (void)touchButton2 {
    [self setOrientationMode];
    NSString *filePath = [self screenShotFilePath];
    [[NCSDKManager getSharedInstance] presentArticlePostViewControllerWithType:kGLArticlePostTypeImage filePath:filePath];
}
- (void)touchButton3 {
    [self setOrientationMode];
    UIAlertView *alert = [[UIAlertView alloc] initWithTitle:nil message:@"ShortCut" delegate:self cancelButtonTitle:@"Cancel" otherButtonTitles:nil, nil];
    [alert addButtonWithTitle:[NSString stringWithFormat:@"Toggle Translate Slider"]];
    [alert show];
}

#pragma mark - NCSDKManagerDelegate
- (void)ncSDKViewDidLoad {
    NSLog(@"ncSDKViewDidLoad");
}
- (void)ncSDKViewDidUnLoad {
    NSLog(@"ncSDKViewDidUnLoad");
}
- (void)ncSDKJoinedCafeMember {
    NSLog(@"Complete To Join Cafe");
}
- (void)ncSDKPostedArticleAtMenu:(NSInteger)menuId
                attachImageCount:(NSInteger)imageCount
                attachVideoCount:(NSInteger)videoCount {
    NSLog(@"Complete To Write, Menu ID[%@]", @(menuId));
}
- (void)ncSDKPostedCommentAtArticle:(NSInteger)articleId {
    NSLog(@"Complete Comment, Article ID[%@]", @(articleId));
}
- (void)ncSDKRequestScreenShot {
    
}
- (void)ncSDKDidVoteAtArticle:(NSInteger)articleId {
    NSLog(@"Complete to vote, Article ID[%@]", @(articleId));
}

- (void)ncSDKWidgetPostArticleWithImage {
    NSString *filePath = [self screenShotFilePath];
    [[NCSDKManager getSharedInstance] presentArticlePostViewControllerWithType:kGLArticlePostTypeImage filePath:filePath];
}
- (void)ncSDKWidgetSuccessVideoRecord {
    [[NCSDKManager getSharedInstance] presentArticlePostViewController];
}
- (void)ncSDKAppSchemeBanner:(NSString *)appScheme {
    NSLog(@"ncSDKAppSchemeBanner %@", appScheme);
}
- (void)ncSDKDidEndWatchingLiveSeconds:(NSInteger)seconds {
    NSLog(@"ncSDKDidEndWatchingLiveSeconds %@", @(seconds));
}
#pragma mark - UIAlertViewDelegate
static BOOL kSliderToggle = NO;
- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex {
    NSString *title = [alertView buttonTitleAtIndex:buttonIndex];
    if ([title hasPrefix:@"Toggle"]) {
        kSliderToggle = !kSliderToggle;
        [[NCSDKManager getSharedInstance] disableTransparentSlider:kSliderToggle];
    }
    
}

- (NSString *)screenShotFilePath {
    UIView *targetScreen = self.view;
    
    UIGraphicsBeginImageContextWithOptions(targetScreen.bounds.size, targetScreen.opaque, 0.0);
    [targetScreen.layer renderInContext:UIGraphicsGetCurrentContext()];
    UIImage *screengrab = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    NSData *pngData = UIImagePNGRepresentation(screengrab);
    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentsPath = [paths objectAtIndex:0];
    NSString *filePath = [documentsPath stringByAppendingPathComponent:@"GLAttachImage.png"];
    [pngData writeToFile:filePath atomically:YES];
    
    return filePath;
}
@end
